File: BestTimetoBuyandSellStockIII2.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        local = [0]*3
        cur = [0]*3
        for i in range(1, len(prices)):
            for j in range(2, 0, -1):
                local[j] = max(local[j]+prices[i]-prices[i-1], cur[j-1]+max(0, prices[i]-prices[i-1]))
                cur[j] = max(cur[j], local[j])
        return cur[2]
        
if __name__=='__main__':
    prices = list(input())
    solution = Solution()
    print(solution.maxProfit(prices))

File: BestTimetoBuyandSellStockIII2.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.empty()) return 0;
        int len=prices.size();
        vector<int> local(3, 0), global(3, 0);
        for(int i=1;i<len;i++){
            for(int j=2;j>=1;j--){
                local[j]=max(global[j-1]+max(0, prices[i]-prices[i-1]), local[j]+prices[i]-prices[i-1]);
                global[j]=max(global[j], local[j]);
            }
        }
        
        return global[2];
    }
    
    int max(int a, int b){
        return a>b?a:b;
    }
};

int main(){
    int n;
    cin>>n;
    
    vector<int> prices;
    for(int i=0;i<n;i++){
        int num;
        cin>>num;
        prices.push_back(num);
    }
    
    Solution *solution=new Solution();
    cout<<solution->maxProfit(prices);
    
    return 0;
}
